source/server/hot_restarting_parent.cc

```diff
@@ -0,0 +1,130 @@
+      continue;
+    }
+    switch (wrapped_request->request().request_case()) {
+    case HotRestartMessage::Request::kShutdownAdmin: {
+      sendHotRestartMessage(child_address_, internal_->shutdownAdmin());
+      break;
+    }
+
+    case HotRestartMessage::Request::kPassListenSocket: {
+      sendHotRestartMessage(child_address_,
+                            internal_->getListenSocketsForChild(wrapped_request->request()));
+      break;
+    }
+
+    case HotRestartMessage::Request::kStats: {
+      HotRestartMessage wrapped_reply;
+      internal_->exportStatsToChild(wrapped_reply.mutable_reply()->mutable_stats());
+      sendHotRestartMessage(child_address_, wrapped_reply);
+      break;
+    }
+
+    case HotRestartMessage::Request::kDrainListeners: {
+      internal_->drainListeners();
+      break;
+    }
+
+    case HotRestartMessage::Request::kTerminate: {
+      ENVOY_LOG(info, "shutting down due to child request");
+      kill(getpid(), SIGTERM);
+      break;
+    }
+
+    default: {
+      ENVOY_LOG(error, "child sent us an unfamiliar type of HotRestartMessage; ignoring.");
+      HotRestartMessage wrapped_reply;
+      wrapped_reply.set_didnt_recognize_your_last_message(true);
+      sendHotRestartMessage(child_address_, wrapped_reply);
+      break;
+    }
+    }
+  }
+}
+
+void HotRestartingParent::shutdown() { socket_event_.reset(); }
+
+HotRestartMessage HotRestartingParent::Internal::shutdownAdmin() {
+  server_->shutdownAdmin();
+  HotRestartMessage wrapped_reply;
+  wrapped_reply.mutable_reply()->mutable_shutdown_admin()->set_original_start_time_unix_seconds(
+      server_->startTimeFirstEpoch());
+  return wrapped_reply;
+}
+
+HotRestartMessage
+HotRestartingParent::Internal::getListenSocketsForChild(const HotRestartMessage::Request& request) {
+  HotRestartMessage wrapped_reply;
+  wrapped_reply.mutable_reply()->mutable_pass_listen_socket()->set_fd(-1);
+  Network::Address::InstanceConstSharedPtr addr =
+      Network::Utility::resolveUrl(request.pass_listen_socket().address());
+  for (const auto& listener : server_->listenerManager().listeners()) {
+    if (*listener.get().socket().localAddress() == *addr) {
+      wrapped_reply.mutable_reply()->mutable_pass_listen_socket()->set_fd(
+          listener.get().socket().ioHandle().fd());
+      break;
+    }
+  }
+  return wrapped_reply;
+}
+
+// TODO(fredlas) if there are enough stats for stat name length to become an issue, this current
+// implementation can negate the benefit of symbolized stat names by periodically reaching the
+// magnitude of memory usage that they are meant to avoid, since this map holds full-string
+// names. The problem can be solved by splitting the export up over many chunks.
+void HotRestartingParent::Internal::exportStatsToChild(HotRestartMessage::Reply::Stats* stats) {
+  for (const auto& gauge : server_->stats().gauges()) {
+    (*stats->mutable_gauges())[gauge->name()] = gauge->value();
```

Comment:
s/gauge->name()/gauge->statName()/ which will be stable.

Reply:
Per Matt's comment, if this failed due to the instability of gauge->name().c_str(), can we add a unit test that shows that failure?

Reply:
Doesn't work there. Also, this wasn't the place where there was an issue.

As for ConstCharStarHashMap, to me, the header file presents ConstCharStarHashMap as a drop-in replacement for a map with a string key. It seems like that might not be the case. If it is not, it could use better documentation, or ideally some sort of mechanical prevention of incorrect use.